Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -715,14 +721,56 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
       let visitedNode = self.visit(declSyntax)
       skipVisitAnyHandling.remove(Syntax(declSyntax))
 
-      let attributesToRemove = self.macroAttributes(attachedTo: visitedNode).map(\.attributeNode)
-
-      return AttributeRemover(removingWhere: { attributesToRemove.contains($0) }).rewrite(visitedNode)
+      let attributesToRemove = self.macroAttributes(attachedTo: visitedNode)
+      attributesToRemove.forEach { (attribute, spec) in
+        if let index = self.expandedAttributes.firstIndex(where: { expandedAttribute in
+          expandedAttribute.position == attribute.position
+        }) {
+          self.expandedAttributes.remove(at: index)
+        } else {
+          if let macroRole = try? inferAttachedMacroRole(definition: spec.type),
+            isInvalidAttachedMacro(macroRole: macroRole, attachedTo: declSyntax)
```

Comment:
Do we even need to check `isInvalidAttachedMacro` here? I would think that we should diagnose all macro attributes that didn’t actually expand. Or am I missing something?

Reply:
I found out how it diagnose when AccessorMacro is applied to struct in the [compiler](https://github.com/apple/swift/blob/d451c3e4e167cc942fe894039f2741f5c40bd2ee/lib/Sema/TypeCheckAttr.cpp#L4001) and imitated it. Also the diagnostics I'v implemented below is just for attached macros(actually only AccessorMacro).

If we should diagnose all macro attributes that didn't expand, what error messages should we emit, a common error message or the same as what compiler emit? Actually I don't know all the cases that attribute won't expand.

Reply:
I made some changes. I've abstracted a method called `diagnosticForUnexpandedMacro`. Currently, it only implements diagnostics for `AccessorMacro`. The rest can be left for others to implement.